modules/fedimint-ln-common/src/api.rs

```diff
@@ -127,3 +141,52 @@ where
         }
     }
 }
+
+/// Filter out duplicate gateways. This is necessary because different guardians
+/// may have different TTLs for the same gateway, so two
+/// `LightningGatewayAnnouncement`s representing the same gateway registration
+/// may not be equal.
+fn filter_duplicate_gateways(
+    gateways: Vec<LightningGatewayAnnouncement>,
+) -> Vec<LightningGatewayAnnouncement> {
+    let gateways_by_gateway_id =
+        gateways
+            .into_iter()
+            .fold(std::collections::HashMap::new(), |mut acc, gateway| {
+                acc.entry(gateway.info.gateway_id)
+                    .or_insert_with(Vec::new)
+                    .push(gateway);
+                acc
+            });
+
+    // For each gateway, we may have multiple announcements with different settings
+    // and/or TTLs. We want to filter out duplicates in a way that doesn't allow a
+    // malicious guardian to override the caller's view of the gateways by
+    // returning a gateway with a shorter TTL. Instead, if we receive multiple
+    // announcements for the same gateway ID, we only filter out announcements
+    // that have the same settings, keeping the one with the longest TTL.
+    gateways_by_gateway_id
+        .into_values()
+        .flat_map(|announcements| {
+            let mut gateways: HashMap<LightningGateway, Duration> = HashMap::new();
+            for announcement in announcements {
+                let ttl = announcement.ttl;
+                let gateway = announcement.info.clone();
+                // Only insert if the TTL is longer than the one we already have
+                gateways
+                    .entry(gateway)
+                    .and_modify(|t| {
+                        if ttl > *t {
+                            *t = ttl;
+                        }
+                    })
+                    .or_insert(ttl);
+            }
+
+            gateways.into_iter().map(|a| LightningGatewayAnnouncement {
```

Comment:
Nit: use `announcement` instead of just `a`

Reply:
Done.